File: bindings/web5_uniffi_wrapper/src/crypto/jwk.rs
use crate::errors::Result;
use web5::crypto::jwk::Jwk as InnerJwk;

pub struct Jwk(pub InnerJwk);

impl Jwk {
    pub fn new(data: InnerJwk) -> Self {
        Self(data)
    }

    pub fn get_data(&self) -> InnerJwk {
        self.0.clone()
    }

    pub fn compute_thumbprint(&self) -> Result<String> {
        let thumbprint = self.0.compute_thumbprint()?;
        Ok(thumbprint)
    }
}

File: bound/golangv2/crypto/jwk.go
package crypto

import (
	"web5/web5c"
)

type JWK struct {
	ALG string `json:"alg,omitempty"`
	KTY string `json:"kty,omitempty"`
	CRV string `json:"crv,omitempty"`
	D   string `json:"d,omitempty"`
	X   string `json:"x,omitempty"`
	Y   string `json:"y,omitempty"`
}

func (j JWK) ComputeThumbprint() (string, error) {
	cJwk := web5c.NewCJwk(j.ALG, j.KTY, j.CRV, j.D, j.X, j.Y)
	defer web5c.FreeCJwk(cJwk)

	thumbprint := web5c.CJwkComputeThumbprint(cJwk)
	return thumbprint, nil
}

func NewJWKFromCJwk(cJwk *web5c.CJwk) *JWK {
	return &JWK{
		ALG: cJwk.GetALG(),
		KTY: cJwk.GetKTY(),
		CRV: cJwk.GetCRV(),
		D:   cJwk.GetD(),
		X:   cJwk.GetX(),
		Y:   cJwk.GetY(),
	}
}
